zerver/lib/create_user.py

```diff
@@ -65,11 +65,29 @@ def copy_default_settings(
     copy_onboarding_steps(settings_source, target_profile)
 
 
+def get_dummy_email_address_for_display_regex(realm: Realm) -> str:
+    """
+    Returns a regex that matches the format of dummy email addresses we
+    generate for the .email of users with limit email_address_visibility.
+
+    The reason we need a regex is that we want something that we can use both
+    for generating the dummy email addresses and recognizing them together with extraction
+    of the user ID.
+    """
+
+    # We can't directly have (\d+) in the username passed to Address, because it gets
+    # mutated by the underlying logic for escaping special characters.
+    # So we use a trick by using $ as a placeholder which will be preserved, and then
+    # replace it with (\d+) to obtain our intended regex.
+    address_template = Address(username="user$", domain=get_fake_email_domain(realm.host)).addr_spec
+    regex = address_template.replace("$", r"(\d+)", 1)
```

Comment:
I think we probably need to do something to protect against the possibility that `get_fake_email_domain(realm.host)` contains regex special characters. @andersk do you have a recommendation here?

Reply:
```suggestion
    regex = re.escape(address_template).replace(r"\$", r"(\d+)", 1)
```